### PR TITLE
BaSP-TG-14: Project controller (getById-delete-update)

### PR DESCRIPTION
**Pull request by cami-diaz**
1. Create project schema.
2. Create routes for projects.
3. Create controllers for projects (getById-delete-update).
4. Apply validation for update.
5. Update postman collection.

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -40,10 +40,10 @@ npm run lint:fix
 
 ### Members
```
Comment (ignacioaristo) [inline Readme.md:41]: This belong to another ticket, please remove it from here and create another branch for this [ticket](https://trello.com/c/3rqBy5mP/2-update-readme).

File: src/models/Projects.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// eslint-disable-next-line import/no-unresolved
```
Comment (ignacioaristo) [inline src/models/Projects.js:1]: Remove this rule.

File: src/routes/projects.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import express from 'express';
+import projects from '../controllers/projects';
+import projectsValidation from '../validations/projects';
+
+const router = express.Router();
+
+router
+  .get('/getById/:id', projects.getProjectById)
```
Comment (ignacioaristo) [inline src/routes/projects.js:8]: Remember that we have changed this, now we assume with the `.get` we are getting by id, so we have to to only `.get(/:id)`

File: src/routes/projects.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  .put('/update/:id', projectsValidation.validateCreation, projects.updateProject)
```
Comment (ignacioaristo) [inline src/routes/projects.js:9]: The same here

File: src/routes/projects.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  .delete('/delete/:id', projects.deleteProject);
```
Comment (ignacioaristo) [inline src/routes/projects.js:10]: and here

File: src/validations/projects.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    clientName: Joi.string().min(3).required(),
+    employee: Joi.array().items(employeeValidation),
+  });
+
+  const validation = projectValidation.validate(req.body);
+
+  if (validation.error) {
+    return res.status(400).json({
+      message: `There was an error: ${validation.error.details[0].message}`,
+      data: undefined,
```
Comment (ignacioaristo) [inline src/validations/projects.js:22]: data in not necessary when you are showing the error.

File: src/validations/projects.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      error: true,
```
Comment (ignacioaristo) [inline src/validations/projects.js:23]: This is not necessary too.

File: src/controllers/projects.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
+import Projects from '../models/Projects';
+
+const getProjectById = async (req, res) => {
+  try {
+    if (!req.params.id) {
+      return res.status(400).json({
+        msg: 'ID was not provided',
+        data: undefined,
```
Comment (ignacioaristo) [inline src/controllers/projects.js:8]: is not necessary retrieve the data with undefined.

File: src/controllers/projects.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        error: true,
```
Comment (ignacioaristo) [inline src/controllers/projects.js:9]: show a boolean value for the error is not necessary for now. Remove all.

File: src/controllers/projects.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
+import Projects from '../models/Projects';
+
+const getProjectById = async (req, res) => {
+  try {
+    if (!req.params.id) {
```
Comment (ignacioaristo) [inline src/controllers/projects.js:5]: is this working correctly?  How do i get this error?

File: src/controllers/projects.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+      });
+    }
+    return res.status(400).json({
+      msg: 'Cannot find project with this ID',
+      data: undefined,
+      error: true,
+    });
+  } catch (error) {
+    return res.status(500).json({
+      msg: 'There was an error',
```
Comment (ignacioaristo) [inline src/controllers/projects.js:27]: I think we can handle the message error better, we can do something like `msg: 'There was an error: ${error}',`

File: src/controllers/projects.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      data: error,
```
Comment (ignacioaristo) [inline src/controllers/projects.js:28]: If we do the previous comment, this is not necessary.

File: src/controllers/projects.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    const projectToDelete = await Projects.findByIdAndDelete({ _id: req.params.id });
+    if (projectToDelete) {
+      return res.status(200).json({
+        msg: 'Project delete succesfully',
+        data: projectToDelete,
+        error: false,
+      });
+    }
+    return res.status(404).json({
+      msg: 'Project was not found',
```
Comment (ignacioaristo) [inline src/controllers/projects.js:52]: Same error handling.

File: src/controllers/projects.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+      data: undefined,
+      error: true,
+    });
+  } catch (error) {
+    return res.status(500).json({
+      msg: 'There was an error',
```
Comment (ignacioaristo) [inline src/controllers/projects.js:93]: Same here

**Review by ignacioaristo — CHANGES_REQUESTED** [review]
Remember to update the Postman collection with the endpoints of projects.

File: src/controllers/projects.js (hunk 10 above)
Comment (cami-diaz) [reply src/controllers/projects.js:5]: I remove it because it was and old code that I forgot to remove.

File: src/controllers/projects.js [hunk 15]
```diff
@@ -0,0 +1,77 @@
+import Projects from '../models/Projects';
+
+const getProjectById = async (req, res) => {
+  try {
+    const project = await Projects.findById({ _id: req.params.id });
```
Comment (JulianFloresDev) [inline src/controllers/projects.js:5]: Hi Cami!! I think you confused the methods of 'findById' with 'findOne'. findOne requires an object as a parameter; while 'findById' requires a string equivalent to an '_id' within the object to be found.

You could try passing the parameter 'req.params.id'.

https://mongoosejs.com/docs/api/model.html#model_Model-findById:~:text=Finds%20a%20single,_id%3A%20null%20%7D).

File: src/controllers/projects.js [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    return res.status(500).json({
+      msg: `There was an error: ${error}`,
+      error: true,
+    });
+  }
+};
+
+const deleteProject = async (req, res) => {
+  try {
+    const projectToDelete = await Projects.findByIdAndDelete({ _id: req.params.id });
```
Comment (JulianFloresDev) [inline src/controllers/projects.js:27]: Same as line 5.

https://mongoosejs.com/docs/api/model.html#model_Model-findByIdAndDelete

File: src/controllers/projects.js [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    return res.status(500).json({
+      msg: `There was an error: ${error}`,
+      error: true,
+    });
+  }
+};
+
+const updateProject = async (req, res) => {
+  try {
+    const projectToUpdate = await Projects.findByIdAndUpdate(
```
Comment (JulianFloresDev) [inline src/controllers/projects.js:49]: Same here! Try to send a number as param in 'findByIdAndUpdate'

File: src/validations/projects.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    role: Joi.string().valid('DEV', 'QA', 'PM', 'TL').required(),
+    rate: Joi.number().required(),
+  });
+  const projectValidation = Joi.object({
+    name: Joi.string().min(3).required(),
+    description: Joi.string().min(3).required(),
+    startDate: Joi.date().required(),
+    endDate: Joi.date().required(),
+    clientName: Joi.string().min(3).required(),
+    employee: Joi.array().items(employeeValidation),
```
Comment (JulianFloresDev) [inline src/validations/projects.js:14]: Would it be possible to send an empty array here?

Like:

'employee': [ ],

I suggest using something like the following

https://joi.dev/api/?v=17.6.1#general-usage:~:text=%2C%20array.ref-,array.min(limit),-Specifies%20the%20minimum

File: src/controllers/projects.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+const deleteProject = async (req, res) => {
+  try {
+    const projectToDelete = await Projects.findByIdAndDelete({ _id: req.params.id });
+    if (projectToDelete) {
+      return res.status(200).json({
+        msg: 'Project delete succesfully',
+        data: projectToDelete,
+        error: false,
```
Comment (ignacioaristo) [inline src/controllers/projects.js:32]: Remove this prop, i think is not necessary for now.
Do it in all the code.

File: src/controllers/projects.js [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    const project = await Projects.findById({ _id: req.params.id });
+    if (project) {
+      return res.status(200).json({
+        msg: 'Project found succesfully',
+        data: project,
+      });
+    }
+    return res.status(400).json({
+      msg: `Cannot find project with ID: ${req.params.id}`,
+      error: true,
```
Comment (ignacioaristo) [inline src/controllers/projects.js:14]: remove all this errors props.

**Review by ignacioaristo — CHANGES_REQUESTED** [review]

**Review by ignacioaristo — APPROVED** [review]

**Pull request merged**